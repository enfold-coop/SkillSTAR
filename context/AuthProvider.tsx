import React, { useState } from "react";
import AsyncStorage from "@react-native-community/async-storage";

type User = null | { username: string };

const AuthContext = React.createContext<{
	user: User;
	login: () => void;
	logout: () => void;
}>({
	user: null,
	login: () => {},
	logout: () => {},
});

interface AuthProviderProps {}

export const AuthProvider: React.FC<AuthProviderProps> = ({ children }) => {
	const [user, setUser] = useState<User>(null);
	return (
		<AuthContext.Provider
			value={{
				user,
				login: () => {
					setUser(user);
					AsyncStorage.setItem("user", JSON.stringify(user));
				},
				logout: () => {
					setUser(null);
					AsyncStorage.removeItem("user");
				},
			}}
		>
			{children}
		</AuthContext.Provider>
	);
};
