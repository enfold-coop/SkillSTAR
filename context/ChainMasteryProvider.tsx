import React, { createContext, useReducer } from 'react';
import {
  ChainMasteryContextDispatch,
  ChainMasteryContextDispatchAction,
  ChainMasteryProviderProps,
  ChainMasteryProviderState,
} from '../types/ChainMasteryProvider';

const initialState: ChainMasteryProviderState = {
  chainMastery: undefined,
};

const ChainMasteryStateContext = createContext<ChainMasteryProviderState | undefined>(undefined);
const ChainMasteryDispatchContext = createContext<ChainMasteryContextDispatch | undefined>(undefined);
const reducer = (state: any, action: ChainMasteryContextDispatchAction) => {
  const actionType = action.type;
  let newState;

  switch (actionType) {
    case 'chainMastery':
      newState = { ...state, chainMastery: action.payload };
      break;
    default:
      throw new Error(`Unhandled action type: ${actionType}`);
  }

  return newState;
};

const ChainMasteryProvider = ({ children }: ChainMasteryProviderProps): JSX.Element => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <ChainMasteryStateContext.Provider value={state}>
      <ChainMasteryDispatchContext.Provider value={dispatch}>{children}</ChainMasteryDispatchContext.Provider>
    </ChainMasteryStateContext.Provider>
  );
};

function useChainMasteryState(): ChainMasteryProviderState {
  const context = React.useContext(ChainMasteryStateContext);

  if (context === undefined) {
    throw new Error('useChainState must be used within a ChainProvider.');
  }

  return context;
}

function useChainMasteryDispatch(): ChainMasteryContextDispatch {
  const context = React.useContext(ChainMasteryDispatchContext);

  if (context === undefined) {
    throw new Error('useChainMasteryDispatch must be used within a ChainMasteryProvider.');
  }

  return context;
}

function useChainMasteryContext(): [ChainMasteryProviderState, ChainMasteryContextDispatch] {
  return [useChainMasteryState(), useChainMasteryDispatch()];
}

export { ChainMasteryProvider, useChainMasteryState, useChainMasteryDispatch, useChainMasteryContext };
