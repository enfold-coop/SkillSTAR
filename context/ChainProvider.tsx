import React, {useReducer} from "react";
import {ChainQuestionnaire} from '../types/CHAIN/ChainQuestionnaire';
import {ChainSession, ChainSessionType} from '../types/CHAIN/ChainSession';
import {ADD_CURR_SESSION_NMBR, ADD_SESSION, ADD_SESSION_TYPE, ADD_USER_DATA,} from "./constants/actions";

type ChainProviderProps = {
  dispatch?: any;
  sessionType: string;
  session?: ChainSession;
  userData?: ChainQuestionnaire;
  currSessionNmbr: number;
  state?: any;
};

const initialState: ChainProviderProps = {
  sessionType: ChainSessionType.probe,
  currSessionNmbr: 0,
};

const store = React.createContext<ChainProviderProps>(initialState);

const {Provider} = store;

const ChainProvider: React.FC = ({children}) => {
  const [state, dispatch] = useReducer((state, action) => {
    switch (action.type) {
      case ADD_SESSION_TYPE:
        return {...state, sessionType: action.payload};
      case ADD_SESSION:
        return {...state, session: action.payload};
      case ADD_USER_DATA:
        return {...state, userData: action.payload};
      case ADD_CURR_SESSION_NMBR:
        return {...state, currSessionNmbr: action.payload};
      default:
        throw new Error();
    }
  }, []);

  return <Provider value={{state, dispatch}}>{children}</Provider>;
};

export {store, ChainProvider};
/**
 * - initialize Session Context Provider
 * - setting/getting state in Context API
 *
 */
