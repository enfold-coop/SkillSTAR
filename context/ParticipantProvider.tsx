import React, { createContext, useReducer } from 'react';
import {
  ParticipantContextDispatch,
  ParticipantContextDispatchAction,
  ParticipantProviderProps,
  ParticipantProviderState,
} from '../types/ParticipantProvider';

const initialState: ParticipantProviderState = {
  participant: undefined,
};

const ParticipantStateContext = createContext<ParticipantProviderState | undefined>(undefined);
const ParticipantDispatchContext = createContext<ParticipantContextDispatch | undefined>(undefined);
const reducer = (state: any, action: ParticipantContextDispatchAction) => {
  const actionType = action.type;
  let newState;

  switch (actionType) {
    case 'participant':
      newState = { ...state, participant: action.payload };
      break;
    default:
      throw new Error(`Unhandled action type: ${actionType}`);
  }

  return newState;
};

const ParticipantProvider = ({ children }: ParticipantProviderProps): JSX.Element => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <ParticipantStateContext.Provider value={state}>
      <ParticipantDispatchContext.Provider value={dispatch}>{children}</ParticipantDispatchContext.Provider>
    </ParticipantStateContext.Provider>
  );
};

function useParticipantState(): ParticipantProviderState {
  const context = React.useContext(ParticipantStateContext);

  if (context === undefined) {
    throw new Error('useChainState must be used within a ChainProvider.');
  }

  return context;
}

function useParticipantDispatch(): ParticipantContextDispatch {
  const context = React.useContext(ParticipantDispatchContext);

  if (context === undefined) {
    throw new Error('useParticipantDispatch must be used within a ParticipantProvider.');
  }

  return context;
}

function useParticipantContext(): [ParticipantProviderState, ParticipantContextDispatch] {
  return [useParticipantState(), useParticipantDispatch()];
}

export { ParticipantProvider, useParticipantState, useParticipantDispatch, useParticipantContext };
