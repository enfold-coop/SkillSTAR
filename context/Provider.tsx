import React, { useEffect, useState } from "react";
import AsyncStorage from "@react-native-community/async-storage";

type ProviderProps = {};

export const Context = React.createContext<Partial<ProviderProps>>({});

export const Provider: React.FC<ProviderProps> = ({ children }) => {
	let [session, setSession] = useState();
	let [stepAttempt, setStepAttempt] = useState();
	let [authorization, setAuthorization] = useState(false);

	return <Context.Provider value={{}}>{children}</Context.Provider>;
};
