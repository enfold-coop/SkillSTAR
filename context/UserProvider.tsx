import React, { createContext, useReducer } from 'react';
import {
  UserContextDispatch,
  UserContextDispatchAction,
  UserProviderProps,
  UserProviderState,
} from '../types/UserProvider';

const initialState: UserProviderState = {
  user: undefined,
};

const UserStateContext = createContext<UserProviderState | undefined>(undefined);
const UserDispatchContext = createContext<UserContextDispatch | undefined>(undefined);
const reducer = (state: any, action: UserContextDispatchAction) => {
  const actionType = action.type;
  let newState;

  switch (actionType) {
    case 'user':
      newState = { ...state, user: action.payload };
      break;
    default:
      throw new Error(`Unhandled action type: ${actionType}`);
  }

  return newState;
};

const UserProvider = ({ children }: UserProviderProps): JSX.Element => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <UserStateContext.Provider value={state}>
      <UserDispatchContext.Provider value={dispatch}>{children}</UserDispatchContext.Provider>
    </UserStateContext.Provider>
  );
};

function useUserState(): UserProviderState {
  const context = React.useContext(UserStateContext);

  if (context === undefined) {
    throw new Error('useChainState must be used within a ChainProvider.');
  }

  return context;
}

function useUserDispatch(): UserContextDispatch {
  const context = React.useContext(UserDispatchContext);

  if (context === undefined) {
    throw new Error('useUserDispatch must be used within a UserProvider.');
  }

  return context;
}

function useUserContext(): [UserProviderState, UserContextDispatch] {
  return [useUserState(), useUserDispatch()];
}

export { UserProvider, useUserState, useUserDispatch, useUserContext };
