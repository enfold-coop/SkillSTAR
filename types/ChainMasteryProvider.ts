import { ReactNode } from 'react';
import { ChainMastery } from '../services/ChainMastery';

export type ChainMasteryContextDispatchAction = { type: 'chainMastery'; payload: ChainMastery };

export type ChainMasteryContextDispatch = (action: ChainMasteryContextDispatchAction) => void;

export type ChainMasteryProviderProps = { children: ReactNode };

export type ChainMasteryProviderState = { chainMastery: ChainMastery | undefined };
