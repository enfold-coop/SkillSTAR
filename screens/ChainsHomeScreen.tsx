import { useDeviceOrientation } from '@react-native-community/hooks';
import { useNavigation } from '@react-navigation/native';
import React, { FC, useEffect, useState } from 'react';
import { FlatList, ImageBackground, StyleSheet, TouchableOpacity, View } from 'react-native';
import * as Animatable from 'react-native-animatable';
import {
  PROBE_INSTRUCTIONS,
  START_PROBE_SESSION_BTN,
  START_TRAINING_SESSION_BTN,
} from '../components/Chain/chainshome_text_assets/chainshome_text';
import ScorecardListItem from '../components/Chain/ScorecardListItem';
import SessionDataAside from '../components/Chain/SessionDataAside';
import AppHeader from '../components/Header/AppHeader';
import { useChainContext } from '../context/ChainProvider';
import { RootNavProps } from '../navigation/root_types';
import { ApiService } from '../services/ApiService';
import { MasteryAlgo } from '../services/MasteryAlgo';
import { MasteryService } from '../services/MasteryService';
import CustomColors from '../styles/Colors';
import { ChainSession, ChainSessionType } from '../types/CHAIN/ChainSession';
import { ChainData, SkillstarChain } from '../types/CHAIN/SkillstarChain';
import { ChainStepStatus, StepAttempt } from '../types/CHAIN/StepAttempt';

type Props = {
  route: RootNavProps<'ChainsHomeScreen'>;
  navigation: RootNavProps<'ChainsHomeScreen'>;
};

// Chain Home Screen
const ChainsHomeScreen: FC<Props> = props => {
  const [asideContent, setAsideContents] = useState('');
  const [btnText, setBtnText] = useState('Start Session');
  const [orient, setOrient] = useState(false);
  const [sessionNmbr, setSessionNmbr] = useState<number>(0);
  const [type, setType] = useState<string>('type');
  const api = new ApiService();
  const [contextState, contextDispatch] = useChainContext();
  const navigation = useNavigation();
  const { portrait } = useDeviceOrientation();

  const callAlgo = (chainData: SkillstarChain) => {
    console.log('*** callAlgo ***');
    MasteryAlgo.determineStepAttemptPromptLevel(chainData);
  };

  useEffect(() => {
    console.log('*** useEffect 1 ***');
    let isCancelled = false;

    if (contextState.chainData != undefined && !isCancelled) {
      // callAlgo(chainData);
      // TODO: Ask the Mastery Service for...
      //  - the current session
      //  - MasteryInfo for each step

      // For now, just use the last ChainSession available in the ChainData.
      if (contextState.chainData.sessions && contextState.chainData.sessions.length > 0) {
        contextDispatch({
          type: 'session',
          payload: contextState.chainData.sessions[contextState.chainData.sessions.length],
        });
        setSessionNmbr(contextState.chainData.sessions.length);
      } else if (contextState.chainSteps && contextState.chainSteps.length > 0) {
        const newChainSession: ChainSession = {
          step_attempts: contextState.chainSteps.map(s => {
            return {
              chain_step_id: s.id,
              chain_step: s,
              completed: false,
              status: ChainStepStatus.not_complete,
            } as StepAttempt;
          }),
        };
        contextDispatch({ type: 'session', payload: newChainSession });
        setSessionNmbr(1);
      }
    }

    return () => {
      isCancelled = true;
    };
  }, [contextState.chainData]);

  useEffect(() => {
    console.log('*** useEffect 2 ***');
    let isCancelled = false;
    let isLoading = false;

    const _load = async () => {
      isLoading = true; // Block while loading

      if (contextState.participant) {
        console.log('contextState.participant =', contextState.participant.name);
        const newData: SkillstarChain = {
          participant_id: contextState.participant.id,
          sessions: [
            {
              session_type: ChainSessionType.probe,
              step_attempts: [],
            },
          ],
        };
        const newDbData = await api.addChainData(newData);
        if (!isCancelled && newDbData) {
          const newChainData = new ChainData(newDbData);
          contextDispatch({ type: 'chainData', payload: newChainData });
        }
      }
    };

    if (!isLoading) {
      _load().then(() => {
        isLoading = false;
        if (!isCancelled) {
          setOrient(portrait);
        }
      });
    }

    return () => {
      isCancelled = true;
    };
  }, [portrait]);

  /**
   * Mastery Algorithm
   * UTIL function (can be moved to another file)
   *
   * Params: chainData = the entire SkillstarChain
   * Returns one of the following:
   * - An empty probe session, if the user has no attempted sessions yet
   * - The next session the participant should be attempting, if there is one.
   * - An empty probe session, if there are none left to attempt (???)
   */
  const setSessionTypeAndNmbr = (chainData: SkillstarChain) => {
    console.log('*** setSessionTypeAndNmbr ***');
    console.log(chainData.sessions.length);

    // Some of the sessions will be future/not attempted sessions.
    // We want the next session the participant should be attempting.
    const numSessions = chainData.sessions ? chainData.sessions.length : 0;

    // If there are no sessions, return a probe session.

    // Otherwise, return the first un-attempted session OR the last attempted session, if there are no un-attempted sessions?

    const lastSess = numSessions > 0 ? chainData.sessions[numSessions - 1] : null;
    // console.log(lastSess);

    // !! overriding type for dev purposes
    // lastSess.session_type = ChainSessionType.training;

    if (lastSess === null) {
      setSessionNmbr(1);
      setType('probe');

      // Session count (how many sessions attempted)
      // i.e., sessions with attempts. Sessions with no attempts would not be included in this count?
      contextDispatch({ type: 'sessionNumber', payload: 1 });

      // chainData.sessions[i].session_type
      contextDispatch({ type: 'sessionType', payload: 'probe' });
    }
    if (lastSess) {
      if (lastSess.session_type === ChainSessionType.training) {
        setType('training');
        // console.log(chainData.sessions.length + 1);

        setSessionNmbr(chainData.sessions.length + 1);
        contextDispatch({ type: 'sessionNumber', payload: sessionNmbr });
        contextDispatch({ type: 'sessionType', payload: 'training' });
        setBtnText(START_TRAINING_SESSION_BTN);
      }
      if (lastSess.session_type === ChainSessionType.probe) {
        // console.log(chainData.sessions.length + 1);
        setType('probe');
        setSessionNmbr(chainData.sessions.length + 1);
        contextDispatch({ type: 'sessionNumber', payload: sessionNmbr });
        contextDispatch({ type: 'sessionType', payload: 'probe' });
        setBtnText(START_PROBE_SESSION_BTN);
        setAsideContents(PROBE_INSTRUCTIONS);
      }
    }
  };

  const navToProbeOrTraining = () => {
    navigation.navigate('PrepareMaterialsScreen', { chainSession: contextState.session });
  };

  const key = contextState.chainData ? contextState.chainData.participant_id : -1;
  const { chainData, chainSteps, session } = contextState;
  const chainSessionId = session && session.id !== undefined ? session.id : -1;

  return (
    <ImageBackground
      key={'chains_home_sreen_' + key}
      source={require('../assets/images/sunrise-muted.jpg')}
      resizeMode={'cover'}
      style={styles.bkgrdImage}
    >
      <View style={portrait ? styles.container : styles.landscapeContainer}>
        <AppHeader name='Chains Home' />
        {chainSteps && chainData !== undefined && (
          <View style={styles.listContainer}>
            <SessionDataAside asideContent={asideContent} />
            <FlatList
              style={styles.list}
              data={chainSteps}
              keyExtractor={chainStep => 'scorecard_chain_step_' + chainStep.id.toString()}
              renderItem={({ item, index, separators }) => (
                <ScorecardListItem
                  chainStep={item}
                  stepAttempt={chainData.getStep(chainSessionId, item.id)}
                  masteryInfo={MasteryService.getMasteryInfoForStep(chainData, item.id)}
                />
              )}
            />
          </View>
        )}
        <TouchableOpacity
          style={[styles.startSessionBtn, { marginBottom: 0 }]}
          onPress={() => {
            navToProbeOrTraining();
          }}
        >
          <Animatable.Text animation='bounceIn' duration={2000} style={styles.btnText}>
            {btnText}
          </Animatable.Text>
        </TouchableOpacity>
      </View>
    </ImageBackground>
  );
};

const styles = StyleSheet.create({
  bkgrdImage: {
    flex: 1,
  },
  container: {
    flex: 1,
    margin: 0,
    justifyContent: 'space-between',
    alignContent: 'flex-start',
    padding: 10,
    paddingBottom: 80,
  },
  landscapeContainer: {
    flex: 1,
    margin: 0,
    justifyContent: 'flex-start',
    alignContent: 'flex-start',
    padding: 10,
    paddingBottom: 80,
  },
  title: {
    fontSize: 30,
    fontWeight: 'bold',
    paddingLeft: 20,
    alignSelf: 'flex-start',
  },
  separator: {
    marginVertical: 30,
    height: 1,
  },
  listContainer: {
    height: '90%',
    flexDirection: 'row',
    justifyContent: 'space-between',
    backgroundColor: 'rgba(255, 255, 255,0.4)',
    padding: 5,
    margin: 5,
    marginTop: 12,
  },
  list: {
    margin: 5,
    marginBottom: 4,
    padding: 5,
    paddingBottom: 30,
  },
  startSessionBtn: {
    width: '90%',
    alignSelf: 'center',
    margin: 10,
    // marginBottom: 20,
    borderRadius: 10,
    paddingVertical: 10,
    backgroundColor: CustomColors.uva.orange,
  },
  btnText: {
    textAlign: 'center',
    color: CustomColors.uva.white,
    fontSize: 32,
    fontWeight: '500',
  },
});

export default ChainsHomeScreen;
