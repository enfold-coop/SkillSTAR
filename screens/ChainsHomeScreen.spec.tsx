import React from 'react';
import renderer from 'react-test-renderer';
import { wait } from '../_util/testing/wait';
import { ChainMasteryProvider } from '../context/ChainMasteryProvider';
import ChainsHomeScreen from './ChainsHomeScreen';

const mockedNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  const navModule = jest.requireActual('@react-navigation/native');
  return {
    ...navModule,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

describe('ChainsHomeScreen', () => {
  afterEach(async () => {
    await wait();
  });

  it('renders correctly', async () => {
    const tree = await renderer
      .create(
        <ChainMasteryProvider>
          <ChainsHomeScreen />
        </ChainMasteryProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
