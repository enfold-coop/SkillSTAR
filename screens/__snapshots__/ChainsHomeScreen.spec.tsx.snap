// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChainsHomeScreen renders correctly 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Image
    resizeMode="cover"
    source={1}
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": undefined,
          "width": undefined,
        },
        undefined,
      ]
    }
  />
  <View
    style={
      Object {
        "alignContent": "flex-start",
        "flex": 1,
        "justifyContent": "space-between",
        "margin": 0,
        "padding": 10,
        "paddingBottom": 80,
      }
    }
  >
    <View
      style={
        Object {
          "alignSelf": "center",
          "backgroundColor": "transparent",
          "borderBottomColor": "#E57200",
          "borderBottomWidth": 3,
          "color": "#FEFEFE",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={
          Object {
            "alignContent": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "padding": 10,
          }
        }
      >
        <Image
          source={1}
          style={
            Object {
              "height": 30,
              "width": 30,
            }
          }
        />
        <Text
          style={
            Object {
              "color": "#232D4B",
              "fontSize": 20,
              "fontWeight": "800",
              "paddingLeft": 20,
            }
          }
        >
          Chains Home
        </Text>
        <View
          style={
            Object {
              "alignItems": "flex-end",
              "flex": 1,
              "flexDirection": "column",
              "height": 60,
              "justifyContent": "center",
              "margin": -16,
              "padding": 0,
            }
          }
        >
          <Text
            style={
              Object {
                "marginRight": 100,
              }
            }
          >
            Loading...
          </Text>
        </View>
      </View>
    </View>
    <View>
      <View
        accessibilityRole="progressbar"
        accessible={true}
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "height": 24,
              "opacity": 1,
              "width": 24,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "bottom": 0,
                "justifyContent": "center",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <View
              style={
                Object {
                  "height": 24,
                  "transform": Array [
                    Object {
                      "rotate": "45deg",
                    },
                  ],
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 12,
                    "overflow": "hidden",
                    "width": 24,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "height": 24,
                      "transform": Array [
                        Object {
                          "translateY": 0,
                        },
                        Object {
                          "rotate": "-165deg",
                        },
                      ],
                      "width": 24,
                    }
                  }
                >
                  <View
                    collapsable={false}
                    style={
                      Object {
                        "height": 12,
                        "overflow": "hidden",
                        "width": 24,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "borderColor": "#6F7EA0",
                          "borderRadius": 12,
                          "borderWidth": 2.4,
                          "height": 24,
                          "width": 24,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "bottom": 0,
                "justifyContent": "center",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <View
              style={
                Object {
                  "height": 24,
                  "transform": Array [
                    Object {
                      "rotate": "45deg",
                    },
                  ],
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 12,
                    "overflow": "hidden",
                    "top": 12,
                    "width": 24,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "height": 24,
                      "transform": Array [
                        Object {
                          "translateY": -12,
                        },
                        Object {
                          "rotate": "345deg",
                        },
                      ],
                      "width": 24,
                    }
                  }
                >
                  <View
                    collapsable={false}
                    style={
                      Object {
                        "height": 12,
                        "overflow": "hidden",
                        "width": 24,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "borderColor": "#6F7EA0",
                          "borderRadius": 12,
                          "borderWidth": 2.4,
                          "height": 24,
                          "width": 24,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View>
      <View
        accessibilityRole="progressbar"
        accessible={true}
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "height": 24,
              "opacity": 1,
              "width": 24,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "bottom": 0,
                "justifyContent": "center",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <View
              style={
                Object {
                  "height": 24,
                  "transform": Array [
                    Object {
                      "rotate": "45deg",
                    },
                  ],
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 12,
                    "overflow": "hidden",
                    "width": 24,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "height": 24,
                      "transform": Array [
                        Object {
                          "translateY": 0,
                        },
                        Object {
                          "rotate": "-165deg",
                        },
                      ],
                      "width": 24,
                    }
                  }
                >
                  <View
                    collapsable={false}
                    style={
                      Object {
                        "height": 12,
                        "overflow": "hidden",
                        "width": 24,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "borderColor": "#6F7EA0",
                          "borderRadius": 12,
                          "borderWidth": 2.4,
                          "height": 24,
                          "width": 24,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "bottom": 0,
                "justifyContent": "center",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <View
              style={
                Object {
                  "height": 24,
                  "transform": Array [
                    Object {
                      "rotate": "45deg",
                    },
                  ],
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 12,
                    "overflow": "hidden",
                    "top": 12,
                    "width": 24,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "height": 24,
                      "transform": Array [
                        Object {
                          "translateY": -12,
                        },
                        Object {
                          "rotate": "345deg",
                        },
                      ],
                      "width": 24,
                    }
                  }
                >
                  <View
                    collapsable={false}
                    style={
                      Object {
                        "height": 12,
                        "overflow": "hidden",
                        "width": 24,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "borderColor": "#6F7EA0",
                          "borderRadius": 12,
                          "borderWidth": 2.4,
                          "height": 24,
                          "width": 24,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
